### ISSUE TITLE
Commands to MQ Training with VSGN

### ISSUE THREAD

**Issue by JunweiLiang** [showlab/EgoVLP#1]
Hi, thanks for releasing the code!

Could you provide some instructions on how to run VSGN training with EgoVLP features (hyper-parameters, learning rate, etc.)? Thanks!

Junwei

**Comment by QinghongLin** [comment]
Hello Junwei,

Thanks for your interest for our work,
I will update the instruction and related details for MQ next,

Thank you for your patience!

**Comment by QinghongLin** [comment]
Hi Junwei, 

I have uploaded the video features for MQ tasks to G drive: [train&val](https://drive.google.com/file/d/1-HEUCdyfNX7CBZhz40yiyTr7to_p7wUi/view) / [test](https://drive.google.com/file/d/1-JmezY3eIkHKJ1JBA_AA8QWBoY3W2HpS/view), so that you can download it directly.
What you need to do is replace the input features with our features.
and I have attached our config of the best VSGN model in here [config.txt](https://github.com/showlab/EgoVLP/files/9368053/results.txt).

Please try it out and let us know if you have new results.

**Comment by JunweiLiang** [comment]
I have downloaded the features but they seem to be a single file. Are they a single pickle binary with dictionary keys? How to read them and map them to the videos (for example, slowfast8x8_r101_k400/ has 9645 *.pt files each corresponds to a video)?

Thanks.

**Comment by QinghongLin** [comment]
There is a gz file, after unzipping it (I unzip it on my mac), you will see a document that contains multiple `*.pt`.
e.g., `0a8f6747-7f79-4176-85ca-f5ec01a15435.pt`, this pt file corresponding to the **video features of the clip: `0a8f6747-7f79-4176-85ca-f5ec01a15435`**.

The clip information is provided by the MQ metadata, i.e., clip xxx come from the video yyy with start time t1 and end time t2.

**Comment by JunweiLiang** [comment]
I see. The file you provided on Google drive is a .tar.gz file, and I extract it with ```tar -zxf``` and got 2034 *.pt file for the train/val part. Will try them.

**Comment by JunweiLiang** [comment]
So ```0a8f6747-7f79-4176-85ca-f5ec01a15435``` is the clip ID instead of video ID? Could you provide the feature files of the whole video as the VSGN baseline? They read the feature of the whole video and then cut the corresponding clip (see [here](https://github.com/EGO4D/episodic-memory/blob/main/MQ/Utils/dataset.py#L98)). To follow [your instructions](https://github.com/showlab/EgoVLP#mq) I would need this video-level features.

Thanks.

**Comment by QinghongLin** [comment]
Yes, it is the clip ID. And sorry, I am currently unable to provide video-level features, a solution is to rewrite the data loader so that supports clip features as input.

**Comment by srama2512** [comment]
@QinghongLin - Thanks for providing the clip features. I tried training the VSGN model using the Ego4D episodic-memory codebase instructions. But I'm not able to reproduce the val results from the paper. The numbers are quite a bit lower than the paper results (2nd row vs. 3rd row in the figure below). 
<img width="1021" alt="image" src="https://user-images.githubusercontent.com/31301440/190316423-0445e87e-d252-40fe-b8dd-320acb31620a.png">

Here is the training command I used. Note: I modified the data loader to use clip features instead of video features.
```
 python Train.py \
     --use_xGPN \
     --is_train true \
     --dataset ego4d \
     --feature_path data/egovlp_feats_official \
     --checkpoint_path checkpoints/ \
     --tb_dir tb/ \
     --batch_size 24 \
     --train_lr 0.00005 \
     --use_clip_features true \
     --input_feat_dim 256 \
     --num_epoch 100
```

**Comment by QinghongLin** [comment]
Hi, @srama2512 
I am sorry that I have no bandwidth to check the details right now.
I upload the MQ codebase we used, which contains relevant details for you to reference.
Future I will re-organize the code and instructions of these downstream tasks.

[MQ.zip](https://github.com/showlab/EgoVLP/files/9596196/MQ.zip)

**Comment by srama2512** [comment]
@QinghongLin  - Thanks for sharing your code and the hyperparameters. I was able to obtain a similar performance. It turns out that there was a bug in the `test_mq.py` feature-extraction code that I used. I modified `test_mq.py` to increase the batch size here to 128. 
https://github.com/showlab/EgoVLP/blob/dc4a60f2dd7fcdc3206ac05d4f452b1c85361ab2/run/test_mq.py#L77-L87

The calculation of `times = data['video'].shape[0] // batch` does not work when video shape is not a multiple of the batch. It gets much worse when we increase the batch, leaving a residual set of all-zero features in the end. After changing that part of the code to the snippet below, it works as expected.
```
if data['video'].shape[0]% batch == 0:
    times = data['video'].shape[0] // batch
else:
    times = data['video'].shape[0] // batch + 1
```
